content/ADR165-resolving-component-numbers.md

```diff
@@ -0,0 +1,155 @@
+---
+layout: adr
+adr: 165 # replace this number for the PR or ISSUE number
+title: Resolving component numbers for the scripting of the SDK
+date: 2023-01-04
+status: Draft
+type: RFC # pick one of these
+spdx-license: CC0-1.0
+authors:
+  - menduz
+---
+
+## Abstract
+
+This RFC describes the non-functional problems to solve in relation to the component numbers of the SDK. It presents synchronization, usability and ergonomic problems behind the technical requirements for the performance. It presents alternatives as a jurney, to add context on the decisions. The document is still not conclusive.
+
+## Context, Reach & Prioritization
+
+Assuming a deterministic and compiled file (`game.js`) and a set of well-known ECS components for rendering (Transform, Gltf, Material, MeshRenderer). It is easy to think about synchronizing either two instances of the `game.js#1` and `game.js#2` or to broadcast visual updates back-and-forth with a renderer. In this case, all components between all actors share a stable numeric identifier (component number) allowing ultra fast synchronization and CRDT semantics, which were hard technical requirements at the design stage.
+
+But there is more than that, beyond the technical requirements, storing the state in a human readable way  is a fundamental use case of the whole product. In a way that can be shared with artists an developers to leverage their collaboration towards a beautiful and polished final experience. And a key element of iterating an experience (whether they use a text or WYSIWYG editor) is evolution of the assets, and that includes the data stored in the components of the entities.
+
+As of the moment of writing this essay, components are identified by a user-defined number pointing to a schema. And that works perfectly for the use cases described in the first paragraph, and from a technical perspective. But it is very short sighted for the big non-functional requirements like enabling evolution, collaboration and reusability/DRY:
+- Component numbers are defined by developers, in code
+- Third party providers (like physics libraries) may use the same component numbers, generating a conflict
+- There is no way to name components for a human-readable or diffable format, i.e. reading `188:[1,1,1],189:[1,1,1]` is not as good as `position:{x:1,y:1,z:1},scale:{x:1,y:1,z:1}`
+
+And due to the imperative nature of the code, defining components (and their schemas) makes it difficult to statically analyze the environment to extract information from it to create better tools. It would be highly beneficial to "analyze" a project to assess its components rather than "run" the code to extract its information.
```

Comment:
```suggestion
And due to the imperative code nature, defining components (and their schemas) makes it difficult to statically analyze the environment to extract information from it to create better tools. It would be highly beneficial to "analyze" a project to assess its components rather than "run" the code to extract its information.
```